logtail/logtail_test.go

```diff
@@ -23,29 +28,245 @@ func TestFastShutdown(t *testing.T) {
 	l := NewLogger(Config{
 		BaseURL: testServ.URL,
 	}, t.Logf)
-	l.Shutdown(ctx)
+	err := l.Shutdown(ctx)
+	if err != nil {
+		t.Error(err)
+	}
 }
 
-func TestUploadMessages(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
-	uploads := 0
-	testServ := httptest.NewServer(http.HandlerFunc(
+type LogtailTestServer struct {
+	ctx      context.Context
+	cancel   context.CancelFunc
+	uploaded chan int
+	logLines int // number of messages to be received
+	body     []byte
+	bodytext string
+	mu       sync.Mutex
```

Comment:
By convention, mutex fields usually go above all the fields that they protect, and document what fields they protect (usually "mu protects all following fields").

Reply:
No longer need a mutex, the channel handles synchronization needs.

I think the version of this PR earlier this week which had to be reverted was me writing Python code in Go syntax ("what do you mean, multiple threads?")

The first version of this PR was me writing C++ code in Go syntax to fix it ("Locks! Everywhere!")

Hopefully this is closer to idiomatic Go.